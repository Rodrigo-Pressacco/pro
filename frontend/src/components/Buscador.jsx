import React, { useState,useEffect } from "react";
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import "../styles/Buscador.css";
import axios from 'axios';
import url from "../utils/datosInicio.json"

// los details es cambiado por el array de lugares

const Buscador = (props) => {

    //destino seleccionado
    const [datos, setDatos] = useState([null, false]);
    const [destinoAEleccion, setDestinoAEleccion] = useState("");
    const [destinoAEleccionShow, setDestinoAEleccionShow] = useState(false);

    const [dateRange, setDateRange] = useState([null, null]);
    const [startDate, endDate] = dateRange;

    useEffect(()=>{
        axios.get(`http://${url.hostUrl}:8080/ciudades/todos`)
        .then(res => {
            const persons = res.data;
            if (datos[1]===false) {
                setDatos([persons, true])                  
            }
        })
    })
    

    //metodo a cambiar cuando entren las categorias
    let arrayDeLugaresFiltrados = null;
    if(datos[1]){
        arrayDeLugaresFiltrados = datos[0].filter(lugar => lugar.nombre.toLocaleLowerCase()
                                                            .includes(destinoAEleccion.toLocaleLowerCase())
        )
    }

    //Setea el destino a eleccion y lo muestra
    const handleChange = e => {
        setDestinoAEleccion(e.target.value);
        if (e.target.value === "") {
            setDestinoAEleccionShow(false);
        }
        else {
            setDestinoAEleccionShow(true);
        }
    }

    //accionar de las opciones de ayuda
    const handleClickDeLaAyudaBuscadorDestino = e => {
        e.preventDefault()
        setDestinoAEleccion(e.target.innerText)
        setDestinoAEleccionShow(false)
        document.getElementById("buscadorDestino").value = e.target.innerText;
    }

    //Crea la lista de Opciones de ayuda o autocompletado
    function searchList() {
        if (destinoAEleccionShow) {
            return (
                <ul className="buscador-ul">
                    {datos[1]?arrayDeLugaresFiltrados.map((element, i) =>
                        <li onClick={handleClickDeLaAyudaBuscadorDestino} className="buscador-li" key={element.nombre + i}><img src="../images/vectorDeDestino.png" alt="vectorDeDestino" />               {element.nombre}, {element.pais}</li>
                    ):<></>}
                </ul>
            )
        }
    }


    //Evento a modificar cuando el buscar tenga que hacer algo
    function handleClickButton(destino) {
        document.getElementById("buscadorDestino").value = "";
        setDestinoAEleccion("")
        setDateRange([null, null])
        props.onSelect(destino)

        if(startDate== null||endDate == null){
            axios.get(`http://${url.hostUrl}:8080/productos/buscarPorCiudad/${destinoAEleccion.split(",")[0]}`)
            .then(res => {
                const persons = res.data;
                props.onSelect2(persons)
            })
        }else if(destinoAEleccion==""){
            axios.get(`http://${url.hostUrl}:8080/productos/buscarPorCiudadFecha/${startDate.toISOString().substring(0,10)}/${endDate.toISOString().substring(0,10)}`)
            .then(res => {
                const persons = res.data;
                props.onSelect2(persons)
            })
        }else{
            axios.get(`http://${url.hostUrl}:8080/productos/buscarPorCiudadFecha/${startDate.toISOString().substring(0,10)}/${endDate.toISOString().substring(0,10)}?ciudad=${destinoAEleccion.split(",")[0]}`)
            .then(res => {
                const persons = res.data;
                props.onSelect2(persons)
            }) 
        }

        

    }
    

    return (
        <div className="Buscador">
            <h1>Buscar ofertas en hoteles, casas y mucho más</h1>
            <div className="div-Buscador">
                <div className="div-buscarLugar">
                    <input type="search" placeholder="  ¿A donde vamos?" className="form-buscadorDestino"
                        onChange={handleChange} name="buscadorDestino" id="buscadorDestino" />
                    {searchList()}

                </div>
                <div className="div-calendario">
                    <DatePicker
                        selectsRange={true}
                        startDate={startDate}
                        endDate={endDate}

                        monthsShown={2}
                        dateFormat=" dd 'de' MMM. yyyy"
                        className="form-calendarDate"
                        placeholderText="   Check in - Check out"
                        onChange={(update) => setDateRange(update)}
                        minDate={new Date()}
                        shouldCloseOnSelect={false}
                        //excludeDates={[new Date(), subDays(new Date(), 1)]} asi se excluyen

                    />
                </div>

                <button className="form-button" onClick={()=>handleClickButton(destinoAEleccion)}>Buscar</button>
            </div>
        </div>

    )
}

export default Buscador;