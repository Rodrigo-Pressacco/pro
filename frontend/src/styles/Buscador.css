@import url("index.css");

.Buscador {
    background-color: var(--third-color);
    display: flex;
    flex-direction: column;
    align-items: center;
}

.Buscador h1 {
    padding: 20px;
    text-align: center;
    color: #F3F1ED;
    font-size: 36px;
    font-weight: bold;
}

.div-Buscador {
    display: flex;
    justify-content: space-around;
    align-items: flex-start;
    width: 75%;
}

.div-buscarLugar{
    width: 39%;
    position: relative;
    
}


.div-calendario {
    width: 39%;
}

.form-buscadorDestino {
    width: 100%;
    margin: 3px 9px 20px 0px;
    padding-left: 20px;
    height: 44px;    
    border-radius: 5px;    

    font-style: normal;
    font-weight: 500;
    font-size: 14px;
    line-height: 16px;
    border: none;
}


.form-buscadorDestino::placeholder,
.form-calendarDate::placeholder{
    font-family: "Font Awesome 5 Free";
    font-weight: 900;
}

.form-buscadorDestino:focus,
.form-calendarDate:focus{
    outline: none;
}

.form-calendarDate{
    width: 100%;
    height: 44px;
    border-radius: 5px;
    padding-left: 20px;
    margin: 3px 0px 20px 0px;

    font-style: normal;
    font-weight: 500;
    font-size: 14px;
    line-height: 16px;
    border: none;

}



.form-button{
    background-color: var(--main-color);
    font-style: normal;
    font-weight: bold;
    font-size: 16px;
    line-height: 19px;    
    color: #FFFFFF;
    
    text-align: center;
    width: 19%;
    margin: 3px 0px 20px 0px;
    height: 44px;
    border: none;
    border-radius: 5px;
}

.form-button:hover{
    background-color: white;
    color: var(--main-color);
    border: 3px solid var(--main-color);
    transition: all .2s ease-in-out;
}

.buscador-ul{
    position: absolute;
    background-color: white;
    z-index: 1;
    top: 60px;
    width: 100%;
    border: #F3F1ED solid 1px;
    box-sizing: border-box;
    box-shadow: 0px 4px 4px rgb(0 0 0 / 25%);
    border-radius: 0px 0px 5px 5px;
}

.buscador-li{
    padding: 25px 0 21px 25px;
    font-weight: bold;
    line-height: 19px;
    border-bottom: 1px solid var(--main-color);
    display: flex;
    align-items: center;
}

.buscador-li img{
    padding: 0 10px 0;
}

/*DatePicker*/

.react-datepicker__day--in-range,
.react-datepicker__day--keyboard-selected{
    background-color: var(--main-color);
    border-radius: 50%;
}

.div-calendario .react-datepicker{
    width: 600px;
    height: 306px;
    margin-top: -20px;
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    background: #FFFFFF;
    border: none;
    box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
    border-radius: 5px;

    background: linear-gradient(90deg, 
                            rgba(255,255,255,1) 49.9% ,
                            rgba(168,168,168,1) 50% ,
                            rgba(255,255,255,1) 50.1%) ;
}

.div-calendario .react-datepicker__header{
    border-bottom: unset;
    background-color: #FFFFFF;
}

.div-calendario .react-datepicker__current-month{
    margin-bottom: 10px;
}

.div-calendario .react-datepicker__navigation{
    top: 10px;
}

.div-calendario .react-datepicker__navigation--next {
    right: 10px;
}

.div-calendario .react-datepicker__navigation--previous{
    left: 10px;
}

.div-calendario .react-datepicker__day,
.div-calendario .react-datepicker__day-name{
    margin: 5px 4px;
}



@media (max-width: 800px){
    .Buscador h1 {
        padding: 20px;
        text-align: center;
        color: #F3F1ED;
        font-size: 30px;
        font-weight: bold;
    }

    .div-Buscador{
        width: 90%;
    }

    .div-calendario {
        width: 39%;
    }

    .react-datepicker-wrapper {
        width: 100%;
    }

    .div-calendario .react-datepicker{
        width: 590px;
    }


}

@media (max-width: 485px){
    .Buscador h1 {
        padding: 20px;
        text-align: center;
        color: #F3F1ED;
        font-size: 30px;
        font-weight: bold;
    }

    .div-Buscador {
        display: flex;
        justify-content: space-around;
        align-items: flex-start;
        width: 90%;
        flex-direction: column;
    }

    .div-buscarLugar {
        width: 100%;
    }

    .div-calendario {
        width: 100%;
    }

    .form-button {
        width: 100%;
    }
    
    .div-calendario .react-datepicker{
        width: 414px;
        height: 550px;
        flex-direction: column;
        background: #FFFFFF;
    }

    .div-calendario .react-datepicker__day,
    .div-calendario .react-datepicker__day-name{
        margin: 4px 10px;
    }

}