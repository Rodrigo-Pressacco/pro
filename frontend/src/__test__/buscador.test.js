import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import Buscador from '../components/Buscador'

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { createSerializer } from 'enzyme-to-json';
Enzyme.configure({ adapter: new Adapter() });
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));

describe('<Buscador />', () => {
    let buscador
    let setDatos = jest.fn();
    let setDestinoAEleccion = jest.fn(); 
    let setDestinoAEleccionShow = jest.fn(); 
    let useEffect = jest.fn();

    beforeEach(() => {
        buscador = shallow(<Buscador    setDestinoAEleccion={setDestinoAEleccion} 
                                        setDestinoAEleccionShow={setDestinoAEleccionShow}
                                        setDatos={setDatos}
                                        useEffect={useEffect} />)
    });

    it("Deberia mostrar <Buscador/> correctamente", () => {
        expect(buscador).toMatchSnapshot();
    });

    it("Deberia retornar 1 etiqueta h1", () => {
        let node = buscador.find("h1")
        expect(node.length).toEqual(1)
    });

    it("Deberia retornar 1 etiqueta button", () => {
        let node = buscador.find("button")
        expect(node.length).toEqual(1)
    });

})