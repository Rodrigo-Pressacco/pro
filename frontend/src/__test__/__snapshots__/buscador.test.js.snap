// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Buscador /> Deberia mostrar <Buscador/> correctamente 1`] = `
<div
  className="Buscador"
>
  <h1>
    Buscar ofertas en hoteles, casas y mucho más
  </h1>
  <div
    className="div-Buscador"
  >
    <div
      className="div-buscarLugar"
    >
      <input
        className="form-buscadorDestino"
        id="buscadorDestino"
        name="buscadorDestino"
        onChange={[Function]}
        placeholder="  ¿A donde vamos?"
        type="search"
      />
    </div>
    <div
      className="div-calendario"
    >
      <r
        allowSameDay={false}
        className="form-calendarDate"
        customTimeInput={null}
        dateFormat=" dd 'de' MMM. yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        endDate={null}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        minDate={2021-11-24T22:39:36.061Z}
        monthsShown={2}
        nextMonthAriaLabel="Next Month"
        nextMonthButtonLabel="Next Month"
        nextYearAriaLabel="Next Year"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        placeholderText="   Check in - Check out"
        preventOpenOnFocus={false}
        previousMonthAriaLabel="Previous Month"
        previousMonthButtonLabel="Previous Month"
        previousYearAriaLabel="Previous Year"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selectsRange={true}
        shouldCloseOnSelect={false}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearPicker={false}
        startDate={null}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
    </div>
    <button
      className="form-button"
      onClick={[Function]}
    >
      Buscar
    </button>
  </div>
</div>
`;
